File: src/pratice/binarysearch/기타레슨2343.java
package pratice.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class 기타레슨2343 {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static int N, M;
    public static int[] A;
    public static void main(String[] args) throws IOException {
        input();
        pro();
    }

    private static void pro() {
        int L = 0;
        int R = Integer.MAX_VALUE;
        int res = R;
        while(L <= R){
            int mid = (L + R) / 2;
            if(isPossible(mid)){
                res = mid;
                R = mid - 1;
            }else{
                L = mid + 1;
            }
        }
        System.out.println(res);
    }

    private static boolean isPossible(int mid) {
        int cnt = 0;
        long sum = 0;
        for (int i = 0; i < N ; i++) {
            if(A[i] > mid) return false;
            if(sum + A[i] <= mid){
                sum += A[i];
            }else{
                cnt++;
                sum = A[i];
            }
        }
        return cnt < M;
    }

    private static void input() throws IOException {
        String[] split = br.readLine().split(" ");
        N = Integer.parseInt(split[0]);
        M = Integer.parseInt(split[1]);
        A = new int[N];
        for (int i = 0; i < N; i++) {
            A[i] = Integer.parseInt(br.readLine());
        }
    }
}

File: src/pratice/binarysearch/기타레슨2343.kt
package pratice.binarysearch

import kotlin.math.max
/** 정렬하면 틀리는 문제*/
fun main() {
    val (N, M) = readln().split(" ").map { it.toInt() }
    val arr = readln().split(" ").map { it.toInt() }.toMutableList()
    lowerBound(arr, M)
}

private fun lowerBound(arr: MutableList<Int>, m: Int) {
    var L = arr[0]
    var R = 1000000000
    var res = 0
    for (num in arr) L = max(num, L)
    while (L <= R) {
        val mid = (L + R) / 2
        if (isAble(arr, m, mid)) {
            res = mid
            R = mid - 1
        } else {
            L = mid + 1
        }
    }
    println(res)
}

private fun isAble(arr: MutableList<Int>, m: Int, mid: Int): Boolean {
    var sum = 0
    var cnt = 0
    for (num in arr) {
        if (sum + num > mid) {
            cnt++
            sum = num
        } else {
            sum += num
        }
    }
    return cnt < m
}
